src/main/java/de/zalando/aruha/nakadi/service/subscription/state/ClosingState.java

```diff
@@ -0,0 +1,146 @@
+package de.zalando.aruha.nakadi.service.subscription.state;
+
+import de.zalando.aruha.nakadi.service.subscription.model.Partition;
+import de.zalando.aruha.nakadi.service.subscription.zk.ZKSubscription;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Stream;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+class ClosingState extends State {
+    private final Map<Partition.PartitionKey, Long> uncommitedOffsets;
+    private final Map<Partition.PartitionKey, ZKSubscription> listeners = new HashMap<>();
+    private final long lastCommitMillis;
+    private ZKSubscription topologyListener;
+    private static final Logger LOG = LoggerFactory.getLogger(ClosingState.class);
+
+    ClosingState(final Map<Partition.PartitionKey, Long> uncommitedOffsets, final long lastCommitMillis) {
+        this.uncommitedOffsets = uncommitedOffsets;
+        this.lastCommitMillis = lastCommitMillis;
+    }
+
+    @Override
+    public void onExit() {
+        try {
+            freePartitions(new HashSet<>(listeners.keySet()));
+        } finally {
+            if (null != topologyListener) {
+                try {
+                    topologyListener.cancel();
+                } finally {
+                    topologyListener = null;
+                }
+            }
+        }
+    }
+
+    @Override
+    public void onEnter() {
+        final long timeToWaitMillis = getParameters().commitTimeoutMillis - (System.currentTimeMillis() - lastCommitMillis);
+        if (timeToWaitMillis > 0) {
+            scheduleTask(() -> {
+                if (isCurrent()) {
+                    switchState(new CleanupState());
+                }
+            }, timeToWaitMillis, TimeUnit.MILLISECONDS);
+            topologyListener = getZk().subscribeForTopologyChanges(() -> addTask(this::onTopologyChanged));
+            reactOnTopologyChange();
+        } else {
+            switchState(new CleanupState());
+        }
+    }
+
+    private void onTopologyChanged() {
+        if (!isCurrent()) {
+            return;
+        }
+        topologyListener.refresh();
+        reactOnTopologyChange();
+    }
+
+    private void reactOnTopologyChange() {
+        final Map<Partition.PartitionKey, Partition> partitions = new HashMap<>();
+        getZk().runLocked(() -> Stream.of(getZk().listPartitions())
+                .filter(p -> getSessionId().equals(p.getSession()))
+                .forEach(p -> partitions.put(p.getKey(), p)));
+        final Set<Partition.PartitionKey> freeRightNow = new HashSet<>();
+        final Set<Partition.PartitionKey> addListeners = new HashSet<>();
+        for (final Partition p : partitions.values()) {
```

Comment:
There is a problem with this approach. If I'll do like that than I probably can skip one of scenarios (or forget about it), just because they are in different places, and I'm not sure if all the scenarios was covered.
Instead of it - I make a decision about every item that belongs to this session. In this case I will react on every change. After making decision I perform actions. And yes, there must be more comments.